packages/api/src/controllers/user.ts

```diff
@@ -636,9 +653,105 @@ app.post("/token", validatePost("user"), async (req, res) => {
     });
   }
   res.status(201);
-  res.json({ id: user.id, email: user.email, token: token });
+  res.json({ id: user.id, email: user.email, token, refreshToken });
 });
 
+app.post(
+  "/token/refresh",
+  validatePost("refresh-token-payload"),
+  async (req, res) => {
+    const now = Date.now();
+    const { refreshToken } = req.body as RefreshTokenPayload;
+    const refreshTokenObj = await db.jwtRefreshToken.get(refreshToken);
+    if (
+      !refreshTokenObj ||
+      refreshTokenObj.expiresAt < now ||
+      refreshTokenObj.revoked
+    ) {
+      console.log(
+        `Refresh attempt with invalid refresh token=${refreshToken} expiresAt=${refreshTokenObj?.expiresAt} revoked=${refreshTokenObj?.revoked}`
+      );
+      return res.status(401).json({ errors: ["invalid refresh token"] });
+    }
+
+    if (now - refreshTokenObj.lastSeen < req.config.jwtAccessTokenTtl / 2) {
+      console.log(
+        `Revoking token due to potential malicious use of refreshToken=${refreshToken}`
+      );
+      await db.jwtRefreshToken.update(refreshToken, {
+        revoked: true,
+      });
+      return res
+        .status(401)
+        .json({ errors: ["refresh token has already been used too recently"] });
+    }
+
+    const user = await db.user.get(refreshTokenObj.userId);
+    if (user.suspended) {
+      return res.status(403).json({ errors: ["user is suspended"] });
+    }
+
+    const token = signUserJwt(user.id, req.config);
+
+    let newRefreshToken: string;
+    const fullTokenTtl = refreshTokenObj.expiresAt - refreshTokenObj.createdAt;
+    if (
+      refreshTokenObj.expiresAt - now <
+      fullTokenTtl * REFRESH_TOKEN_REFRESH_THRESHOLD
+    ) {
+      // issue a new token and revoke the old one if it's close to expiring
+      await db.jwtRefreshToken.update(refreshToken, {
+        lastSeen: now,
+        revoked: true,
+      });
+
+      newRefreshToken = uuid();
```

Comment:
Shouldn't we first create a new refresh token and then revoke the old one? 

Otherwise, what happens, if we revoked the old one, but failed to create a new one?

Reply:
Yeah I moved these around a bit... I thought it would be worse if we created a new refresh token and then failed to revoke the old one, then there's 2 refresh tokens out there for the same session. I'd rather have the user logout and try again in this rare scenario (and If we think about it, if we have any failure here the user will get a status error and won't get even an access token, let alone a refresh token)